ORBIT API fails for CustomArraySystemDesign

recent changes to play nicely with WOMBAT only work if ArraySystemDesign (and not a custom one) is being used!

see: https://github.com/NLRWindSystems/WISDEM/blob/0d97725155b54ff8ec29fb835cd1bf20a5cc1308/wisdem/orbit/orbit_api.py#L787

so slow... sorry.
```bash
>>> pytest -s test/ard/unit/cost/test_orbit_wrap.py
================================================================= test session starts ==================================================================
platform darwin -- Python 3.14.3, pytest-9.0.2, pluggy-1.6.0
rootdir: /Users/cfrontin/codes/Ard
configfile: pyproject.toml
plugins: cov-7.1.0, anyio-4.13.0
collected 3 items                                                                                                                                      

test/ard/unit/cost/test_orbit_wrap.py ORBIT library intialized at '/Users/cfrontin/codes/Ard/case_files/working/ORBIT_library'
.FF

======================================================================= FAILURES =======================================================================
______________________________________________________ TestORBITNoApproxBranch.test_baseline_farm ______________________________________________________

self = 'orbit.orbit.orbit' <class ORBITWisdemDetail>, fname = 'compute', protect_inputs = True, protect_outputs = False, protect_residuals = False

    @contextmanager
    def _call_user_function(self, fname, protect_inputs=True,
                            protect_outputs=False, protect_residuals=False):
        """
        Context manager that wraps a call to a user defined function.
    
        Protect any vectors that should not be modified to help prevent user error
        and add information about the system to any errors that don't have it already.
    
        Parameters
        ----------
        fname : str
            Name of the user defined function.
        protect_inputs : bool
            If True, then set the inputs vector to be read only
        protect_outputs : bool
            If True, then set the outputs vector to be read only
        protect_residuals : bool
            If True, then set the residuals vector to be read only
        """
        self._inputs.read_only = protect_inputs
        self._outputs.read_only = protect_outputs
        self._residuals.read_only = protect_residuals
    
        try:
>           yield

../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/system.py:2952: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/explicitcomponent.py:342: in _compute_wrapper
    self.compute(self._inputs, self._outputs,
ard/cost/orbit_wrap.py:381: in compute
    super().compute(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = 'orbit.orbit.orbit' <class ORBITWisdemDetail>, inputs = <openmdao.vectors.default_vector.DefaultVector object at 0x33499d710>
outputs = <openmdao.vectors.default_vector.DefaultVector object at 0x33499d9b0>
discrete_inputs = <openmdao.core.component._DictValues object at 0x33487d730>
discrete_outputs = <openmdao.core.component._DictValues object at 0x33487d790>

    def compute(self, inputs, outputs, discrete_inputs, discrete_outputs):
        """Creates and runs the project, then gathers the results."""
        quiet_flag = self.options["quiet"]
    
        config = self.compile_orbit_config_file(
            inputs,
            outputs,
            discrete_inputs,
            discrete_outputs,
        )
    
        project = ProjectManager(config)
        if quiet_flag:
            with HiddenPrints():
                project.run()
        else:
            project.run()
    
        # The ORBIT version of total_capex includes turbine capex, so we do our own sum of
        # the parts here that wisdem doesn't account for
        capacity_kW = 1e3 * inputs["turbine_rating"] * discrete_inputs["number_of_turbines"]
        outputs["bos_capex"] = project.bos_capex
        outputs["soft_capex"] = project.soft_capex
        outputs["project_capex"] = project.project_capex
        outputs["total_capex"] = project.bos_capex + project.soft_capex + project.project_capex
        outputs["total_capex_kW"] = outputs["total_capex"] / capacity_kW
        outputs["installation_time"] = project.installation_time
        outputs["installation_capex"] = project.installation_capex
        outputs["capacity"] = project.capacity
>       discrete_outputs["layout"] = project.phases["ArraySystemDesign"].create_layout_df()
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       KeyError: 'ArraySystemDesign'

../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/wisdem/orbit/orbit_api.py:808: KeyError

During handling of the above exception, another exception occurred:

self = <cost.test_orbit_wrap.TestORBITNoApproxBranch object at 0x3318ab890>, subtests = <_pytest.subtests.Subtests object at 0x333ee07d0>

    def test_baseline_farm(self, subtests):
    
        # specify the configuration/specification files to use
        filename_turbine = (
            Path(ard.__file__).parents[1]
            / "examples"
            / "data"
            / "windIO-plant_turbine_IEA-22MW-284m-RWT.yaml"
        ).absolute()  # toolset generalized turbine specification
    
        # set up the modeling options
        windIO_plant = {
            "site": {
                "energy_resource": {
                    "wind_resource": {
                        "shear": 0.2,
                    },
                },
            },
            "wind_farm": {
                "turbine": ard.utils.io.load_yaml(filename_turbine),
                "electrical_substations": [
                    {
                        "electrical_substation": {
                            "coordinates": {
                                "x": [100.0],
                                "y": [100.0],
                            },
                        },
                    }
                ],
            },
        }
        modeling_options = {
            "windIO_plant": windIO_plant,
            "layout": {
                "N_turbines": 7,
                "N_substations": 1,
                "x_turbines": np.linspace(-7.0 * 400.0, 7.0 * 400.0, 7),
                "y_turbines": np.linspace(7.0 * 400.0, -7.0 * 400.0, 7),
            },
            "offshore": True,
            "floating": True,
            "platform": {
                "N_anchors": 3,
                "min_mooring_line_length_m": 500.0,
                "N_anchor_dimensions": 2,
            },
            "site_depth": 50.0,
            "collection": {
                "max_turbines_per_string": 8,
                "model_options": dict(
                    topology="radial",
                    feeder_route="segmented",
                    feeder_limit="unlimited",
                ),
                "solver_name": "highs",
                "solver_options": dict(
                    time_limit=10,
                    mip_gap=0.005,  # TODO ???
                ),
            },
            "costs": {
                "rated_power": 22000000.0,  # (W)
                "num_blades": 3,
                "tower_length": 149.386,  # (m)
                "tower_mass": 1574044.87111,  # (tonne)
                "nacelle_mass": 849143.2357,  # (tonne)
                "blade_mass": 83308.31171,  # (tonne)
                "turbine_capex": 1397.17046735,  # (tonne)
                "site_mean_windspeed": 10.0,  # (m/s)
                "turbine_rated_windspeed": 11.13484394,  # (m/s)
                "commissioning_cost_kW": 44.0,  # (USD/kW)
                "decommissioning_cost_kW": 58.0,  # (USD/kW)
                "plant_substation_distance": 1.0,  # (km)
                "interconnection_distance": 8.5,  # (km)
                "site_distance": 115.0,  # (km)
                "site_distance_to_landfall": 50.0,  # (km)
                "port_cost_per_month": 2000000.0,  # (USD/mo)
                "construction_insurance": 44.0,  # (USD/kW)
                "construction_financing": 183.0,  # (USD/kW)
                "contingency": 316.0,  # (USD/kW)
                "site_auction_price": 100000000.0,  # (USD)
                "site_assessment_cost": 50000000.0,  # (USD)
                "construction_plan_cost": 250000.0,  # (USD)
                "installation_plan_cost": 1000000.0,  # (USD)
                "boem_review_cost": 0.0,  # (USD)
                "transition_piece_mass": 100.0e3,  # (kg)
                "transition_piece_cost": 0.0,  # (USD)
                # # Fixed bottom configuration
                # monopile_mass: 2097.21115974 # (tonne)
                # monopile_cost: 4744119.28172591 # (USD)
                # tcc_per_kW: 1397.17046735 # (USD/kW)
                # opex_per_kW: 110. # (USD/kW)
                # # Floating configuration
                "num_mooring_lines": 3,  # (-)
                "mooring_line_mass": 843225.1875,  # (kg)
                "mooring_line_diameter": 0.225,  # (m)
                "mooring_line_length": 837.0,  # (m)
                "anchor_mass": 0.0,  # (kg)
                "floating_substructure_cost": 11803978.242949858,  # (USD)
            },
        }
    
        # create an OM model and problem
        model = om.Group()
        coll = model.add_subsystem(  # collection component
            "collection",
            ard.collection.OptiwindnetCollection(
                modeling_options=modeling_options,
            ),
            promotes=[
                "x_turbines",
                "y_turbines",
                "x_substations",
                "y_substations",
            ],
        )
    
        orbit = model.add_subsystem(
            "orbit",
            ocost.ORBITDetailedGroup(
                modeling_options=modeling_options,
                approximate_branches=False,
            ),
            promotes=[
                "x_turbines",
                "y_turbines",
                "x_substations",
                "y_substations",
            ],
        )
        model.connect("collection.graph", "orbit.graph")
    
        model.set_input_defaults(
            "x_turbines", modeling_options["layout"]["x_turbines"], units="km"
        )
        model.set_input_defaults(
            "y_turbines", modeling_options["layout"]["y_turbines"], units="km"
        )
        model.set_input_defaults(
            "x_substations",
            [
                li["electrical_substation"]["coordinates"]["x"]
                for li in windIO_plant["wind_farm"]["electrical_substations"]
            ],
            units="km",
        )
        model.set_input_defaults(
            "y_substations",
            [
                li["electrical_substation"]["coordinates"]["y"]
                for li in windIO_plant["wind_farm"]["electrical_substations"]
            ],
            units="km",
        )
    
        prob = om.Problem(model)
        prob.setup()
    
        prob.set_val("x_turbines", modeling_options["layout"]["x_turbines"], units="m")
        prob.set_val("y_turbines", modeling_options["layout"]["y_turbines"], units="m")
    
        prob.set_val(
            "x_substations",
            [
                li["electrical_substation"]["coordinates"]["x"]
                for li in windIO_plant["wind_farm"]["electrical_substations"]
            ],
            units="km",
        )
        prob.set_val(
            "y_substations",
            [
                li["electrical_substation"]["coordinates"]["y"]
                for li in windIO_plant["wind_farm"]["electrical_substations"]
            ],
            units="km",
        )
    
>       prob.run_model()

test/ard/unit/cost/test_orbit_wrap.py:379: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/problem.py:653: in run_model
    self.model.run_solve_nonlinear()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/system.py:5112: in run_solve_nonlinear
    self._solve_nonlinear()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/group.py:2585: in _solve_nonlinear
    self._nonlinear_solver._solve_with_cache_check()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:26: in _solve_with_cache_check
    self.solve()  # don't use caching
    ^^^^^^^^^^^^
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:45: in solve
    self._gs_iter()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/solver.py:937: in _gs_iter
    subsys._solve_nonlinear()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/group.py:2585: in _solve_nonlinear
    self._nonlinear_solver._solve_with_cache_check()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:26: in _solve_with_cache_check
    self.solve()  # don't use caching
    ^^^^^^^^^^^^
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:45: in solve
    self._gs_iter()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/solver.py:937: in _gs_iter
    subsys._solve_nonlinear()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/group.py:2585: in _solve_nonlinear
    self._nonlinear_solver._solve_with_cache_check()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:26: in _solve_with_cache_check
    self.solve()  # don't use caching
    ^^^^^^^^^^^^
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:45: in solve
    self._gs_iter()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/solver.py:937: in _gs_iter
    subsys._solve_nonlinear()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/explicitcomponent.py:371: in _solve_nonlinear
    self._compute_wrapper()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/explicitcomponent.py:325: in _compute_wrapper
    with self._call_user_function('compute'):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../miniforge3/envs/ard-test-env/lib/python3.14/contextlib.py:162: in __exit__
    self.gen.throw(value)
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/system.py:2958: in _call_user_function
    raise err_type(
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/system.py:2952: in _call_user_function
    yield
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/explicitcomponent.py:342: in _compute_wrapper
    self.compute(self._inputs, self._outputs,
ard/cost/orbit_wrap.py:381: in compute
    super().compute(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = 'orbit.orbit.orbit' <class ORBITWisdemDetail>, inputs = <openmdao.vectors.default_vector.DefaultVector object at 0x33499d710>
outputs = <openmdao.vectors.default_vector.DefaultVector object at 0x33499d9b0>
discrete_inputs = <openmdao.core.component._DictValues object at 0x33487d730>
discrete_outputs = <openmdao.core.component._DictValues object at 0x33487d790>

    def compute(self, inputs, outputs, discrete_inputs, discrete_outputs):
        """Creates and runs the project, then gathers the results."""
        quiet_flag = self.options["quiet"]
    
        config = self.compile_orbit_config_file(
            inputs,
            outputs,
            discrete_inputs,
            discrete_outputs,
        )
    
        project = ProjectManager(config)
        if quiet_flag:
            with HiddenPrints():
                project.run()
        else:
            project.run()
    
        # The ORBIT version of total_capex includes turbine capex, so we do our own sum of
        # the parts here that wisdem doesn't account for
        capacity_kW = 1e3 * inputs["turbine_rating"] * discrete_inputs["number_of_turbines"]
        outputs["bos_capex"] = project.bos_capex
        outputs["soft_capex"] = project.soft_capex
        outputs["project_capex"] = project.project_capex
        outputs["total_capex"] = project.bos_capex + project.soft_capex + project.project_capex
        outputs["total_capex_kW"] = outputs["total_capex"] / capacity_kW
        outputs["installation_time"] = project.installation_time
        outputs["installation_capex"] = project.installation_capex
        outputs["capacity"] = project.capacity
>       discrete_outputs["layout"] = project.phases["ArraySystemDesign"].create_layout_df()
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       KeyError: "'orbit.orbit.orbit' <class ORBITWisdemDetail>: Error calling compute(), 'ArraySystemDesign'"

../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/wisdem/orbit/orbit_api.py:808: KeyError
_______________________________________________________ TestORBITApproxBranch.test_baseline_farm _______________________________________________________

self = 'orbit.orbit.orbit' <class ORBITWisdemDetail>, fname = 'compute', protect_inputs = True, protect_outputs = False, protect_residuals = False

    @contextmanager
    def _call_user_function(self, fname, protect_inputs=True,
                            protect_outputs=False, protect_residuals=False):
        """
        Context manager that wraps a call to a user defined function.
    
        Protect any vectors that should not be modified to help prevent user error
        and add information about the system to any errors that don't have it already.
    
        Parameters
        ----------
        fname : str
            Name of the user defined function.
        protect_inputs : bool
            If True, then set the inputs vector to be read only
        protect_outputs : bool
            If True, then set the outputs vector to be read only
        protect_residuals : bool
            If True, then set the residuals vector to be read only
        """
        self._inputs.read_only = protect_inputs
        self._outputs.read_only = protect_outputs
        self._residuals.read_only = protect_residuals
    
        try:
>           yield

../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/system.py:2952: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/explicitcomponent.py:342: in _compute_wrapper
    self.compute(self._inputs, self._outputs,
ard/cost/orbit_wrap.py:381: in compute
    super().compute(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = 'orbit.orbit.orbit' <class ORBITWisdemDetail>, inputs = <openmdao.vectors.default_vector.DefaultVector object at 0x336903310>
outputs = <openmdao.vectors.default_vector.DefaultVector object at 0x3369035b0>
discrete_inputs = <openmdao.core.component._DictValues object at 0x334c23e30>
discrete_outputs = <openmdao.core.component._DictValues object at 0x334c23ed0>

    def compute(self, inputs, outputs, discrete_inputs, discrete_outputs):
        """Creates and runs the project, then gathers the results."""
        quiet_flag = self.options["quiet"]
    
        config = self.compile_orbit_config_file(
            inputs,
            outputs,
            discrete_inputs,
            discrete_outputs,
        )
    
        project = ProjectManager(config)
        if quiet_flag:
            with HiddenPrints():
                project.run()
        else:
            project.run()
    
        # The ORBIT version of total_capex includes turbine capex, so we do our own sum of
        # the parts here that wisdem doesn't account for
        capacity_kW = 1e3 * inputs["turbine_rating"] * discrete_inputs["number_of_turbines"]
        outputs["bos_capex"] = project.bos_capex
        outputs["soft_capex"] = project.soft_capex
        outputs["project_capex"] = project.project_capex
        outputs["total_capex"] = project.bos_capex + project.soft_capex + project.project_capex
        outputs["total_capex_kW"] = outputs["total_capex"] / capacity_kW
        outputs["installation_time"] = project.installation_time
        outputs["installation_capex"] = project.installation_capex
        outputs["capacity"] = project.capacity
>       discrete_outputs["layout"] = project.phases["ArraySystemDesign"].create_layout_df()
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       KeyError: 'ArraySystemDesign'

../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/wisdem/orbit/orbit_api.py:808: KeyError

During handling of the above exception, another exception occurred:

self = <cost.test_orbit_wrap.TestORBITApproxBranch object at 0x3318ab9d0>, subtests = <_pytest.subtests.Subtests object at 0x3358f2990>

    def test_baseline_farm(self, subtests):
    
        self.prob.set_val(
            "x_turbines", self.modeling_options["layout"]["x_turbines"], units="m"
        )
        self.prob.set_val(
            "y_turbines", self.modeling_options["layout"]["y_turbines"], units="m"
        )
    
        self.prob.set_val(
            "x_substations",
            [
                li["electrical_substation"]["coordinates"]["x"]
                for li in self.windIO_plant["wind_farm"]["electrical_substations"]
            ],
            units="km",
        )
        self.prob.set_val(
            "y_substations",
            [
                li["electrical_substation"]["coordinates"]["y"]
                for li in self.windIO_plant["wind_farm"]["electrical_substations"]
            ],
            units="km",
        )
    
>       self.prob.run_model()

test/ard/unit/cost/test_orbit_wrap.py:582: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/problem.py:653: in run_model
    self.model.run_solve_nonlinear()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/system.py:5112: in run_solve_nonlinear
    self._solve_nonlinear()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/group.py:2585: in _solve_nonlinear
    self._nonlinear_solver._solve_with_cache_check()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:26: in _solve_with_cache_check
    self.solve()  # don't use caching
    ^^^^^^^^^^^^
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:45: in solve
    self._gs_iter()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/solver.py:937: in _gs_iter
    subsys._solve_nonlinear()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/group.py:2585: in _solve_nonlinear
    self._nonlinear_solver._solve_with_cache_check()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:26: in _solve_with_cache_check
    self.solve()  # don't use caching
    ^^^^^^^^^^^^
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:45: in solve
    self._gs_iter()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/solver.py:937: in _gs_iter
    subsys._solve_nonlinear()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/group.py:2585: in _solve_nonlinear
    self._nonlinear_solver._solve_with_cache_check()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:26: in _solve_with_cache_check
    self.solve()  # don't use caching
    ^^^^^^^^^^^^
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/nonlinear/nonlinear_runonce.py:45: in solve
    self._gs_iter()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/solvers/solver.py:937: in _gs_iter
    subsys._solve_nonlinear()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/explicitcomponent.py:371: in _solve_nonlinear
    self._compute_wrapper()
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/explicitcomponent.py:325: in _compute_wrapper
    with self._call_user_function('compute'):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../miniforge3/envs/ard-test-env/lib/python3.14/contextlib.py:162: in __exit__
    self.gen.throw(value)
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/system.py:2958: in _call_user_function
    raise err_type(
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/system.py:2952: in _call_user_function
    yield
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/openmdao/core/explicitcomponent.py:342: in _compute_wrapper
    self.compute(self._inputs, self._outputs,
ard/cost/orbit_wrap.py:381: in compute
    super().compute(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = 'orbit.orbit.orbit' <class ORBITWisdemDetail>, inputs = <openmdao.vectors.default_vector.DefaultVector object at 0x336903310>
outputs = <openmdao.vectors.default_vector.DefaultVector object at 0x3369035b0>
discrete_inputs = <openmdao.core.component._DictValues object at 0x334c23e30>
discrete_outputs = <openmdao.core.component._DictValues object at 0x334c23ed0>

    def compute(self, inputs, outputs, discrete_inputs, discrete_outputs):
        """Creates and runs the project, then gathers the results."""
        quiet_flag = self.options["quiet"]
    
        config = self.compile_orbit_config_file(
            inputs,
            outputs,
            discrete_inputs,
            discrete_outputs,
        )
    
        project = ProjectManager(config)
        if quiet_flag:
            with HiddenPrints():
                project.run()
        else:
            project.run()
    
        # The ORBIT version of total_capex includes turbine capex, so we do our own sum of
        # the parts here that wisdem doesn't account for
        capacity_kW = 1e3 * inputs["turbine_rating"] * discrete_inputs["number_of_turbines"]
        outputs["bos_capex"] = project.bos_capex
        outputs["soft_capex"] = project.soft_capex
        outputs["project_capex"] = project.project_capex
        outputs["total_capex"] = project.bos_capex + project.soft_capex + project.project_capex
        outputs["total_capex_kW"] = outputs["total_capex"] / capacity_kW
        outputs["installation_time"] = project.installation_time
        outputs["installation_capex"] = project.installation_capex
        outputs["capacity"] = project.capacity
>       discrete_outputs["layout"] = project.phases["ArraySystemDesign"].create_layout_df()
                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       KeyError: "'orbit.orbit.orbit' <class ORBITWisdemDetail>: Error calling compute(), 'ArraySystemDesign'"

../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/wisdem/orbit/orbit_api.py:808: KeyError
=================================================================== warnings summary ===================================================================
../../miniforge3/envs/ard-test-env/lib/python3.14/site-packages/jsonmerge/resolver.py:2
  DeprecationWarning: /Users/cfrontin/miniforge3/envs/ard-test-env/lib/python3.14/site-packages/jsonmerge/resolver.py:2
  jsonschema.RefResolver is deprecated as of v4.18.0, in favor of the https://github.com/python-jsonschema/referencing library, which provides more compliant referencing behavior as well as more flexible APIs for customization. A future release will remove RefResolver. Please file a feature request (on referencing) if you are missing an API for the kind of customization you need.

test/ard/unit/cost/test_orbit_wrap.py::TestORBITNoApproxBranch::test_baseline_farm
test/ard/unit/cost/test_orbit_wrap.py::TestORBITApproxBranch::test_baseline_farm
  UserWarning: /Users/cfrontin/miniforge3/envs/ard-test-env/lib/python3.14/site-packages/ORBIT/phases/design/array_system_design.py:1100
  Missing data in columns ['cable_length', 'bury_speed']; all values will be calculated.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============================================================== short test summary info ================================================================
FAILED test/ard/unit/cost/test_orbit_wrap.py::TestORBITNoApproxBranch::test_baseline_farm - KeyError: "'orbit.orbit.orbit' <class ORBITWisdemDetail>: Error calling compute(), 'ArraySystemDesign'"
FAILED test/ard/unit/cost/test_orbit_wrap.py::TestORBITApproxBranch::test_baseline_farm - KeyError: "'orbit.orbit.orbit' <class ORBITWisdemDetail>: Error calling compute(), 'ArraySystemDesign'"
======================================================= 2 failed, 1 passed, 3 warnings in 5.91s ========================================================
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -4,7 +4,7 @@ build-backend = "setuptools.build_meta"
 
 [project]
 name = "wisdem"
-version = "4.1.0"
+version = "4.1.1"
 description = "Wind-Plant Integrated System Design & Engineering Model"
 readme = "README.md"
 requires-python = ">=3.9"

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -56,7 +56,7 @@ dependencies = [
   "numpy",
   "openmdao",
   "openpyxl",
-  "orbit-nrel>=1.2.5",
+  "orbit-nrel>=1.2.6",
   "pandas",
   "pydoe3",
   "pyyaml",

--- a/wisdem/__init__.py
+++ b/wisdem/__init__.py
@@ -1,1 +1,5 @@
+from importlib.metadata import version
+
 from wisdem.glue_code.runWISDEM import run_wisdem
+
+__version__ = version("wisdem")

--- a/wisdem/orbit/orbit_api.py
+++ b/wisdem/orbit/orbit_api.py
@@ -805,4 +805,7 @@ class OrbitWisdem(om.ExplicitComponent):
         outputs["installation_time"] = project.installation_time
         outputs["installation_capex"] = project.installation_capex
         outputs["capacity"] = project.capacity
-        discrete_outputs["layout"] = project.phases["ArraySystemDesign"].create_layout_df()
+        if "ArraySystemDesign" in project.phases:
+            discrete_outputs["layout"] = project.phases["ArraySystemDesign"].create_layout_df()
+        elif "CustomArraySystemDesign" in project.phases:
+            discrete_outputs["layout"] = project.phases["CustomArraySystemDesign"].create_layout_df()

